fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/AddDecodeNodeForDictStringRule.java

```diff
@@ -399,6 +377,24 @@ public OptExpression visitPhysicalOlapScan(OptExpression optExpression, DecodeCo
                     context.hasEncoded = true;
                 }
 
+                // rewrite predicate
+                for (Integer columnId : context.tableIdToStringColumnIds.get(tableId)) {
+                    if (context.stringColumnIdToDictColumnIds.containsKey(columnId) &&
+                            scanOperator.getPredicate() != null &&
+                            scanOperator.getPredicate().getUsedColumns().contains(columnId)) {
+                        for (int i = 0; i < predicates.size(); i++) {
+                            ScalarOperator predicate = predicates.get(i);
+                            if (predicate.getUsedColumns().contains(columnId)) {
+                                Preconditions.checkState(
+                                        couldApplyDictOptimize(predicate, context.allStringColumnIds));
+                                final DictMappingRewriter rewriter = new DictMappingRewriter(context);
+                                final ScalarOperator newCallOperator = rewriter.rewrite(predicate.clone());
+                                predicates.set(i, newCallOperator);
+                            }
+                        }
+                    }
+                }
+
```

Comment:
suggest to compute out a ColumnSet from  context.tableIdToStringColumnIds.get(tableId) (line#381), then use iterate predicates, and use the predicate.getUsedColumns.intersect(dictSet).
this code is for nested if-stmts and for-stmts, to hard to read.